quinn-h3/src/headers.rs

```diff
@@ -0,0 +1,47 @@
+use futures::{Async, Future, Poll};
+use quinn_proto::StreamId;
+
+use crate::{
+    connection::ConnectionRef,
+    proto::{frame::HeadersFrame, headers::Header},
+    Error,
+};
+
+pub struct DecodeHeaders {
+    trailers: Option<HeadersFrame>,
+    conn: ConnectionRef,
+    stream_id: StreamId,
+}
+
+impl DecodeHeaders {
+    pub(crate) fn new(trailers: HeadersFrame, conn: ConnectionRef, stream_id: StreamId) -> Self {
+        Self {
+            conn,
+            stream_id,
+            trailers: Some(trailers),
+        }
+    }
+}
+
+impl Future for DecodeHeaders {
+    type Item = Header;
+    type Error = crate::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.trailers {
+            None => Err(crate::Error::Internal("trailers none")),
+            Some(ref frame) => {
+                let result = {
+                    let conn = &mut self.conn.h3.lock().unwrap().inner;
```

Comment:
This whole thing can just be folded into a single expression, which then doesn't need the explicit block?

Reply:
Right. Fixed.